doc/devdocs/types.rst

```diff
@@ -0,0 +1,522 @@
+
+Here ``ub`` is ``Integer``, as specified in the function definition.
+
+The last field of a ``TypeVar`` is ``bound``.  This boolean value
+specifies whether the ``TypeVar`` is defined as one of the function
+parameters.  For example::
+
+   julia> h1(A::Array, b::Real) = 1
+   h1 (generic function with 1 method)
+
+   julia> h2{T<:Real}(A::Array, b::T) = 1
+   h2 (generic function with 1 method)
+
+   julia> h3{T<:Real}(A::Array{T}, b::T) = 1
+   h3 (generic function with 1 method)
+
+   julia> p1 = start(methods(h1)).sig.parameters
+   svec(Array{T,N},Real)
+
+   julia> p2 = start(methods(h2)).sig.parameters
+   svec(Array{T,N},T<:Real)
+
+   julia> p3 = start(methods(h3)).sig.parameters
+   svec(Array{T<:Real,N},T<:Real)
+
+   julia> xdump(p1[1].parameters[1])
+   TypeVar
+     name: Symbol T
+     lb: Union()
+     ub: Any::DataType  <: Any
+     bound: Bool false
+
+   julia> xdump(p3[1].parameters[1])
+   TypeVar
+     name: Symbol T
+     lb: Union()
+     ub: Real::DataType  <: Number
+     bound: Bool true
+
+Note that ``p2`` shows two objects called ``T``, but only one of them
+has the upper bound ``Real``; in contrast, ``p3`` shows both of them
+bounded.  This is because in ``h3``, the same type ``T`` is used in
+both places, whereas for ``h2`` the ``T`` inside the array is simply
+the default symbol used for the first parameter of ``Array``.
+
+One can construct ``TypeVar``\s manually::
+
+   julia> TypeVar(:V, Signed, Real, false)
+   Signed<:V<:Real
+
+There are convenience versions that allow you to omit any of these
+arguments except the ``name`` symbol.
+
+Armed with this information, we can do some sneaky things that reveal
+a lot about how julia does dispatch::
+
+   julia> TV = TypeVar(:T, false)   # bound = false
+   T
+
+   julia> candid{T}(A::Array{T}, x::T) = 0
+   candid (generic function with 1 method)
+
+   julia> @eval sneaky{T}(A::Array{T}, x::$TV) = 1
+   sneaky (generic function with 1 method)
+
+   julia> methods(candid)
+   # 1 method for generic function "candid":
+   candid{T}(A::Array{T,N},x::T) at none:1
+
+   julia> methods(sneaky)
+   # 1 method for generic function "sneaky":
+   sneaky{T}(A::Array{T,N},x::T) at none:2
+
+These therefore print identically, but they have very different behavior::
+
+   julia> candid([1],3.2)
+   ERROR: MethodError: `candid` has no method matching candid(::Array{Int64,1}, ::Float64)
+   Closest candidates are:
+     candid{T}(::Array{T,N}, ::T)
+
+   julia> sneaky([1],3.2)
+   1
+
+To see what's happening, it's helpful to use julia's internal ``jl_``
+function (defined in ``builtins.c``) for display, because it prints
+bound ``TypeVar`` objects with a hash (``#T`` instead of ``T``)::
+
+   julia> jl_(x) = ccall(:jl_, Void, (Any,), x)
+   jl_ (generic function with 1 method)
+
+   julia> jl_(start(methods(candid)))
+   Method(sig=Tuple{Array{#T<:Any, N<:Any}, #T<:Any}, va=false, isstaged=false, tvars=#T<:Any, func=#<function>, invokes=nothing, next=nothing)
```

Comment:
Does anyone know how to prevent these `#` from being interpreted as comment-characters?